src/wp-includes/general-template.php

```diff
@@ -3422,6 +3422,121 @@ function wp_resource_hints() {
+	/**
+	 * Filters domains and URLs for resource preloads.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array  $urls {
+	 *     Array of resources and their attributes, or URLs to print for resource preloads.
+	 *
+	 *     @type array ...$0 {
+	 *         Array of resource attributes.
+	 *
+	 *         @type string $href        URL to include in resource preloads. Required.
+	 *         @type string $as          How the browser should treat the resource
+	 *                                   (`script`, `style`, `image`, `document`, etc).
+	 *         @type string $crossorigin Indicates the CORS policy of the specified resource.
+	 *         @type string $type        Type of the resource (`text/html`, `text/css`, etc).
+	 *         @type string $media       Accepts media types or media queries. Allows responsive preloading.
+	 *         @type string $imagesizes  Responsive source size to the source Set.
+	 *         @type string $imagesrcset Responsive image sources to the source set.
+	 *     }
+	 * }
+	 */
+	$urls = apply_filters( 'wp_preload_links', array() );
+
+	if ( ! is_array( $urls ) ) {
+		return;
+	}
+
+	$unique_urls = array();
+	foreach ( $urls as $url ) {
+		if ( ! is_array( $url ) ) {
+			continue;
+		}
+
+		$atts = $url;
+		if ( isset( $url['href'] ) ) {
+			$url = $url['href'];
+			if ( isset( $unique_urls[ $url ] ) ) {
+				continue;
+			}
+			$unique_urls[ $url ] = $atts;
+			// Media can use imagesrcset and not href.
+		} elseif ( ! isset( $url['href'] )
+				&& ( 'image' === $url['as'] )
+				&& ( isset( $url['imagesrcset'] ) || isset( $url['imagesizes'] ) )
+		) {
+			if ( isset( $unique_urls[ $url['imagesrcset'] ] ) ) {
+				continue;
+			}
+			$unique_urls[ $url['imagesrcset'] ] = $atts;
+		} else {
+			continue;
+		}
+
+		foreach ( $unique_urls as $unique_atts ) {
+			$html = '';
+
+			foreach ( $unique_atts as $attr => $value ) {
+				/**
+				 * Ignore not supported attributes.
+				 */
+				if ( ! is_scalar( $value )
+					|| (
+							! in_array( $attr, array( 'as', 'crossorigin', 'href', 'imagesrcset', 'imagesizes', 'type', 'media' ), true )
+							&& ! is_numeric( $attr )
+					)
+				) {
+
+					continue;
+				}
+
+				// imagesrcset only usable when preloading image, ignore otherwise.
+				if ( ( 'imagesrcset' === $attr ) && ( ! isset( $unique_atts['as'] ) || ( 'image' !== $unique_atts['as'] ) ) ) {
+					continue;
+				}
+
+				// imagesizes only usable when preloading image and imagesrcset present, ignore otherwise.
+				if ( ( 'imagesizes' === $attr )
+					&& ( ! isset( $unique_atts['as'] ) || ( 'image' !== $unique_atts['as'] ) || ! isset( $unique_atts['imagesrcset'] ) )
+				) {
+					continue;
+				}
+
+				$value = ( 'href' === $attr ) ? esc_url( $value, array( 'http', 'https' ) ) : esc_attr( $value );
+
+				if ( ! is_string( $attr ) ) {
+					$html .= " $value";
+				} else {
+					$html .= " $attr='$value'";
+				}
+			}
+		}
+
+		$html = trim( $html );
+
+		echo "<link rel='preload' $html />\n";
```

Comment:
```suggestion
		printf("<link rel='preload' %s />\n", $html);
```

This feels a little safer.